### ISSUE TITLE
WC Subscriptions and WC Pay Subscription state/status should remain in sync (aka WCPay subs inexplicably paused for specific merchant)

### ISSUE THREAD

**Issue by james-allan** [Automattic/woocommerce-payments#4547]
### Description

It's currently possible under some error condition for the subscription record in Stripe to become out of sync with the subscription in WooCommerce. This p2 (pbUcTB-162-p2) outlines a scenario where subscriptions in WC Pay (Stripe) and the corresponding subscription in WC became out of sync and there was no indication to the merchant why these subscriptions stopped renewing. 

### Acceptance criteria

* If a subscription becomes out of sync with this WC Pay counterpart, repair the data. 

⚠️ This issue will need to fleshed out once we have a better idea on what the subscription's source of truth (the WC subscription status or Stripe status) should be. 

### Testing instructions

Given this happens under some error condition you will need to simulate the error by:

1. Purchase a subscription using WC Pay (WC Subscriptions plugin disabled)
2. With WP Console, add the following code snippet to your site. 
3. At this point there needs to be a process to reactivate the subscription in Stripe - ie resync the subscription with Stripe. 

```php
$subscription_id = 570; // Replace this ID with yours.
WC_Payments_Subscriptions::get_subscription_service()->suspend_subscription( wcs_get_subscription( $subscription_id ) ); 
```

**Comment by james-allan** [comment]
If we decide to use Stripe as the source of truth, one approach would be to listen for the [`customer.subscription.updated` event](https://stripe.com/docs/api/events/types#event_types-customer.subscription.updated) and update the WC subscription based on the data sent. The object sent is the whole subscription object so would contain all data needed to keep it synced.

**Comment by james-allan** [comment]
Even if we were to use the [`customer.subscription.updated` event](https://stripe.com/docs/api/events/types#event_types-customer.subscription.updated) I mentioned above, it may still be possible for WC to get out of sync if there's errors that prevent data syncs, database rollbacks, or direct database queries etc. 

We may not be able to avoid all possible data misalignments but one fall back option could be to add an indicator on the edit subscription screen that says the subscription is out of sync, with perhaps a button to resync it.

**Comment by thenbrent** [comment]
> ⚠️ This issue will need to fleshed out once we have a better idea on what the subscription's source of truth (the WC subscription status or Stripe status) should be.

In general WC is the UI for managing the subscriptions, so it should be the source of truth (i.e. there is no dashboard for a merchant or subscriber to manage their subscription on Stripe's side).

That said, I suspect there will be conditions where the subscription is suspended or otherwise has its status changed in Stripe's side that lead to this issue and that we need to address. Do we know what those conditions are? If not, how can we figure out the complete list?

Based on the P2 today, it seems like we still haven't found the source of the original issue which caused subscriptions to be voided on Stripe's side. That would be helpful to identify any corner cases. My hunch so far is that was initiated from WC's side somehow, perhaps through a staging site, and in that case, Stripe would need to be treated as the source of truth to ensure the live site's record of the subscription is still in sync (and can be adjusted accordingly).

**Comment by johndcoy** [comment]
>Do we know what those conditions are? If not, how can we figure out the complete list?

@thenbrent We have an issue in [zd-5542919](https://woothemes.zendesk.com/agent/tickets/5542919) where the [customer's September 12th renewal ](https://dashboard.stripe.com/connect/accounts/acct_1JwaCQ2H65Yyiz4N/customers/cus_L8hgqh1tAsnztN)was canceled in WCPay due to `Cancellation reason: void_invoice`, and what appears to be future renewals:

![Screenshot](https://d.pr/i/jnmeDZ+)
Screenshot link: https://d.pr/i/jnmeDZ

All the while the subscription remains active on WooCommerce with no additional renewals or scheduled pending actions.

![Screenshot](https://d.pr/i/hn7b7O+)
Screenshot link: https://d.pr/i/hn7b7O

![Screenshot](https://d.pr/i/hXYT5O+)
Screenshot link: https://d.pr/i/hXYT5O

![Screenshot](https://d.pr/i/Oi87XY+)
Screenshot link: https://d.pr/i/Oi87XY

I assume the `Cancellation reason: void_invoice` is a result of the payment retry failing after the allowed time or something else - the truth is I have no clue. I would like to see if we can get any insight as to why the retry failed and/or a callback for why the collection was paused.

Whatever complexity that may cause this ( invalid stored cards after switching blog IDs, etc. etc. ) I think that it would be best for there to be a renewal order created in WooCommerce and it checks the status of the subscription in Stripe ( catching collection paused ) giving some insight into actionable items.

Hope that insight is helpful.

**Comment by james-allan** [comment]
> I assume the Cancellation reason: void_invoice is a result of the payment retry failing after the allowed time or something else - the truth is I have no clue

I had a look at the subscription in question and the subscription was suspended in Stripe (pause collection) on Aug 19th at 8:30am UTC. 

<img width="886" alt="Screen Shot 2022-10-05 at 11 11 44 am" src="https://user-images.githubusercontent.com/8490476/193958346-861deb0a-50de-453f-a85f-54a65f7764bd.png">

This is a request that came from the WC Payments server most likely from a request from the store. 

Stripe handle renewals a little differently to us. They still create the scheduled renewal invoice in September, but because the subscription was paused, it's automatically voided - this is what the "pause collection: void" means - while the subscription is paused, void all future invoices. 

Looking at the corresponding subscription on the WC store, there's no indication for why this pause request would have been sent on that date.

> I think that it would be best for there to be a renewal order created in WooCommerce and it checks the status of the subscription in Stripe ( catching collection paused ) giving some insight into actionable items.

FWIW if the payment had failed, an order would have been created in WC giving the merchant and customer an opportunity to correct the payment method etc. In this case it was just purely out of sync issue where the subscription in WC was active but in Stripe was on-hold (paused).

Having said that, when Stripe create the voided invoice, we might be able use that opportunity to check if things aren't in sync etc. 

Other than that, we don't have any specific ideas for how we're going to sync the two systems but we could use a scheduled action similar to the renewal payment one that checks if things are synced on a more regular basis.

**Comment by Brianmitchtay** [comment]
Seeing this issue come up again in 5672639-zd-woothemes, no clear indication for what is causing these subscriptions to change their status to `Collection paused`. There's nothing in the subscription notes that indicate anything. But the API calls setting the subscriptions to `pause collection` are the same as mentioned above.

In the time that I have been looking at the ticket and trying to wrap my head around the issue, 6 more subscriptions have set themselves to `pause collection`.

**Comment by csmcneill** [comment]
Here's how we're currently handling this:

1. WooCommerce Happiness Engineers can re-enable payment collection for affected merchants if they want to continue using WCPay. Alternatively, they can move to WC Subs, but they will have to follow [this guide](https://woocommerce.com/document/payments/subscriptions/upgrading-to-subscriptions/#all-subscriptions).
2. To recoup fees for the missed payment, the merchant can manually send an invoice. They can create a brand new order and use the `Email invoice / order details to customer` option from the Order Actions metabox for a customer to pay for this missing payment: ![XWznLd.png](https://user-images.githubusercontent.com/82999806/199861672-4d02ce4f-3576-460e-960a-c13678783483.png)

**Comment by nicdwilson** [comment]
In 5817776-zen The merchant has had subscriptions pausing multiple times, and has had to ask for a manual restart each time. The merchant is hesitant to undertake the subscription migration route currently outlined in the documentation, because, among other reasons, the price of their subscriptions has changed, and forcing subscribers to repurchase would create complications.

Their level of frustration is peaking and I believe they are likely to churn completely.

This is not the first merchant I have seen put off by the manual migration now described in the documentation.

Previously Ohana was able to migrate WC Pay subscriptions by

- removing wcpay subs meta id (we had a plugin for that)
- resave the subscription with payment method data set
- doublecheck scheduled dates were correct and in place by resaving scheduled next payment plus one minute

This used to be enough to get a sub up and running on tokens already stored. Would the process above still work, rather than forcing subscribers to repurchase?

@Automattic/helix (as requested p1672228585011879-slack-C019K6Z9C8Y)

**Comment by AZStolman** [comment]
It is now four months since this was posted.  Our site has experienced "pausing" three times since July 2022.  Woo is still "investigating"  with no real solution other than continue manual billing missed payments or migrate to the Woo subscriptions plugin.  We too are hesitant to undertake the subscription migration route because the price of our subscriptions has changed, the products have changed and forcing our subscribers to repurchase could likely result in cancellations.

**Comment by Brianmitchtay** [comment]
Seeing a slightly different variation of this theme (subscriptions on the Stripe billing engine not staying in sync) in 5854817-zen, where a merchant has had a subscription continue billing despite them having cancelled it in their wp-admin area. 

This has caused them to churn completely from WooCommerce Payments.

I've asked them for any details they can recall about their process of cancelling it to try and see if we can reproduce.

**Comment by csmcneill** [comment]
@Brianmitchtay 5854817-zen is affected by #5385 (thanks for opening that up!) instead of this one, right?

**Comment by Brianmitchtay** [comment]
@csmcneill Yes technically it's #5385 I'd say. Though it fits with the broader theme of "Subscriptions should stay in sync" for sure!

**Comment by AZStolman** [comment]
Where did the customer “churn” to?  

Assume it was a better place.

**Comment by shevene** [comment]
+1 5970880-zen

**Comment by csmcneill** [comment]
6202573-zen

**Comment by kaushikasomaiya** [comment]
6264031-zen

**Comment by thenbrent** [comment]
@james-allan can this one be closed now that #6344 is merged/shipped?

**Comment by james-allan** [comment]
Yep. This issue was hinting that we have some sort of sync job/process and I'm less inclined to think that's an idea worth pursuing. 

While resyncing regularly may catch some situations where Stripe and WC get out of sync, I think there are far too many unknowns around how Stripe might behave if we updated a Stripe billing subscription mid stream. Especially knowing that prorating is on by default and can do all sorts of wacky things like create small $ invoices after updating payment dates.

Best case scenario this resync wouldn't need to do anything 99.99% of the time. Worst case scenario it leads to unforeseen issues. In all scenarios, it would increase the API surface area that we need to maintain and support.
